### PR TITLE
Lower-friction scheduled queries: Support manual definition of external dependencies

### PR DESCRIPTION
**Pull request by scholtzan**
Depends on https://github.com/mozilla/bigquery-etl/pull/1020

There are a few cases where queries are dependent on tasks scheduled in telemetry-airflow (for example, if data is loaded into tables by a job running on GKE). I think it might be useful to have an option in query `metadata.yaml` files to specify these external dependencies.

### CODE REVIEW

File: sql/telemetry_derived/deviations_v1/metadata.yaml [hunk 1]
```diff
@@ -10,3 +10,8 @@ labels:
   public_bigquery: true
   review_bug: 1624528
   incremental_export: false
+scheduling:
+  dag_name: bqetl_deviations
+  depends_on:
+    - task_id: anomdtct
+      dag_name: public_analysis
```
Comment (scholtzan) [inline sql/telemetry_derived/deviations_v1/metadata.yaml:17]: Dependencies could be listed like this.

File: dags/bqetl_deviations.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+    )
+
+    wait_for_public_analysis_anomdtct = ExternalTaskSensor(
+        task_id="wait_for_public_analysis_anomdtct",
+        external_dag_id="public_analysis",
+        external_task_id="anomdtct",
+        dag=dag,
+    )
+
+    telemetry_derived__deviations__v1.set_upstream(wait_for_public_analysis_anomdtct)
```
Comment (scholtzan) [inline dags/bqetl_deviations.py:46]: ...and would appear as dependencies for scheduled tasks

**Comment by scholtzan** [comment]
Still needs tests. But I wanted to make sure this is something we actually (which I think we do 🤔).

Also most changes come from https://github.com/mozilla/bigquery-etl/pull/1020

**Comment by jklukas** [comment]
I'll review once #1020 is merged and this is rebased.

File: bigquery_etl/query_scheduling/task.py [hunk 3]
```diff
@@ -45,6 +45,20 @@ class UnscheduledTask(Exception):
...
+    Representation of a reference to another task.
+
+    The task can be defined in bigquery-etl or in telemetry-airflow.
+    Uses attrs to simplify the class definition and provide validation.
+    Docs: https://www.attrs.org
+    """
+
+    dag_name: str = attr.ib()
+    task_id: str = attr.ib()
+
```
Comment (scholtzan) [inline bigquery_etl/query_scheduling/task.py:60]: We likely have to add an `execution_delta` option here too.

Comment (jklukas) [reply bigquery_etl/query_scheduling/task.py:60]: Do we also need to perform validation to make sure we don't have duplicate task dependencies listed between the `depends_on` list and automatically inferred dependencies?

File: bigquery_etl/query_scheduling/task.py [hunk 4]
```diff
@@ -205,6 +231,13 @@ def with_dependencies(self, client, dag_collection):
...
             if upstream_task is not None:
-                dependencies.append(upstream_task)
+                # ensure there are no duplicate dependencies
+                # manual dependency definitions overwrite automatically detected ones
+                if not any(
+                    d.dag_name == upstream_task.dag_name
+                    and d.task_id == upstream_task.task_name
+                    for d in self.depends_on
+                ):
+                    dependencies.append(upstream_task)
```
Comment (scholtzan) [inline bigquery_etl/query_scheduling/task.py:241]: Handle duplicate task dependencies listed between the `depends_on` list and automatically inferred dependencies here

**Comment by scholtzan** [comment]
Ready for review.

**Comment by jklukas** [comment]
I'm going to merge this as I want to rebase https://github.com/mozilla/bigquery-etl/pull/1027 on it and get more direct experience with generating a DAG with that AMO stats use case.

**Pull request merged**